### ISSUE TITLE
Visual representation of key in authorisation requests

### ISSUE THREAD

**Issue by gtlewis** [UniversalLogin/UniversalLoginSDK#159]
A malicious gas relayer could substitute the key in an authorisation request with a key they control. Therefore a visual representation of the key (e.g. blockie) should be displayed with pending authorisations (instead of browser / OS / IP). This should also be shown when requesting to connect - so user can visually confirm they match.

**Comment by alexvandesande** [comment]
Thank you. But we should discuss other solutions for the same problem: could you use a PIN code instead, so that a malicious actor cannot send authorizations? Or maybe use something other than blockies (some apps use a 4 emoji check)

**Comment by gtlewis** [comment]
The problem with a PIN code is that the malicious actor could still swap in their own key under the code. Hence why a visual representation of the key works, agree that blockies may not be the best way - 4 emojis sounds promising.

**Comment by alexvandesande** [comment]
Wait, what is the threat model here? Is the malicious attacker in control of the interface code? Because then they can also show the incorrect blockie.

**Comment by gtlewis** [comment]
No, it is rather a malicious gas relayer. Currently pending authorisations are stored in the gas relayers (#157) - so a malicious gas relayer could substitute their own key in the authorisation request.

**Comment by gtlewis** [comment]
No, it is rather a malicious gas relayer. Currently pending authorisations are stored in the gas relayers (#156) - so a malicious gas relayer could substitute their own key in the authorisation request.

**Comment by alexvandesande** [comment]
Valid attack vector, and it's interesting because even if we had pins or other identifying information like IP, the attacker can still use the same.

I wonder if that;'s not a case in which emojis would be more interesting to use:

![emoji](https://user-images.githubusercontent.com/112898/49593211-2a7ce080-f95a-11e8-897a-109a2f4359a0.jpeg)

**Comment by alexvandesande** [comment]
Pinging @drhus for his work on emojicode

**Comment by drhus** [comment]
If I understand correctly each device has its own key, and if you use blockie to visually represent this new device adding, this would create a confusion for the user, as he will see 2 blockies, the one on the top the ID/ENS and now new one the new device confirmation. 

I agree with Alex, a picto-string _or emoji-like__ casual visual checksumming is a superior solution for such use cases, and work the same as One-Time Password combating man-in-the-middle attacks. _I don't have any emperical study to back this claim but [i'm working on one](https://momcode.io/memorygame/) ;)_

Q. How can we implement that with UniversalLogin?

**Emoji:** Like [Radix-emoji](https://github.com/MidnightLightning/radix-emoji), [Ecoji](https://github.com/keith-turner/ecoji) or [Base Emoji](https://github.com/pfrazee/base-emoji) very straightforward base-conversion, further 

1) Non use filtered set of emojies, and there are a lot of very similar Emojis already.
2) Emojis aren't consistent across devices and platforms, and could look [completely different](https://mentalfloss.com/article/516048/22-emojis-look-completely-different-different-phones), 

You can solve that by shipping all Emojies used a couple of hundreds up to 2000+ depending on the approach we're taking.

**Momcode** 
they are basically 16 to 32 easily recognized colorful pictograph like heart, square, star, triangle etc.. like ![screenshot of momcode __ deterministic visual identifier _ test lab v0 6](https://user-images.githubusercontent.com/712919/50014807-1e250300-ffac-11e8-9ca5-35bde7c01385.jpg) 
or 
 ![screenshot of momcode __ deterministic visual identifier _ test lab v0 6 1](https://user-images.githubusercontent.com/712919/50014802-1cf3d600-ffac-11e8-87c4-decbcb066a3c.jpg) 


check out  https://momcode.io/lab/

the downside is that they aren't as beautiful as Emojis, from the other side, but they are consistent across devices as it's only up to 32 pictos on svg come part of the library.

btw there is a Emoji set at [Momcode](https://momcode.io/lab/?mode=HEX&input=493598b106ab9ea7e743ac603ba5498ebf181e4a&list=4Fun%20-%20Emoji-Random) but it's not any better than Ecoji & Base-Emoji above (just smaller set of 256)

**attack, collision and possible values**
using a pool of 1000 emojis as Ecoji, with 4 of them we have 1+ trillion possible combinations. (1000^4)
using a pool of 32 Momcode picto, with 4 of them we have 400+ billion, and with 5 pictos it's 1+ trillion (32*8 colors^5)

_actially the possible combination above isn't the indication for collision probability, due to the birthday paradox. but still this good enough.._

**Comment by gtlewis** [comment]
Pinging @tkali

**Comment by gtlewis** [comment]
The emojis are nice - however thinking more about this:

> I see no reason the user needs to know or understand the key they have on their device, so if it's simply about comparing something with the other

Actually I believe the visual verification does have to be derived from the public address/key - else a malicious gas relayer (or any malicious actor sending an authorisation request) can substitute their own key in the request. E.g. if request is sent to authorise `0x123...` with symbols mouse/surprise/wind/elephant" then a bogus request can be sent to authorise `0x666`... also with the symbols mouse/surprise/wind/elephant. Deriving the symbols from the address/key requesting authorisation prevents this.

Of course it doesn't have to be blockies (the emojis are much nicer...) and user having to type the emjos is sound idea.

**Comment by drhus** [comment]
There is several javascript emoji pickers (emoji keyboards) but this is going to be a pain really, very inconvenient for the user to select that 4 emojis out of thousands of emojis..

even if we use Momcode's [keyboard ](ttps://momcode.io/lab/)(there is a simple keyboard module) it's going to be a pain still to find your emoji out of 256 ones 
![screenshot of momcode __ deterministic visual identifier _ test lab v0 6 3](https://user-images.githubusercontent.com/712919/51493641-1b80af00-1d9d-11e9-9c26-e3389722b982.jpg)

the symbols version is simpler however still a lot 
![screenshot of momcode __ deterministic visual identifier _ test lab v0 6 2](https://user-images.githubusercontent.com/712919/51493629-17549180-1d9d-11e9-8955-ebef189ff2d3.jpg)
ps. this last keyboard can be reduced by 80% by putting colors on tabs 

While complete Momcode/Emojis keyboard can be inconvenient, we can alternatively 

**1)** show the authorization pictos among 3x, allowing the user to select the 4 pictographs from 12 (so the keyboard would be the 12) ex.

![screenshot of momcode __ deterministic visual identifier _ test lab v0 6 4](https://user-images.githubusercontent.com/712919/51494272-3ce29a80-1d9f-11e9-874a-a7f692bf577b.jpg)

**2)** or alternatively showing the same authorization 4 pictos on the remote device but in random order, user only need to sort out the order by dragging them right and left

> Actually I believe the visual verification does have to be derived from the public address/key

I agree, but keeping in mind a malicious gas relayer can use a vanity address generator to generate an address which he control, that has the same 4 authorization picto/emojis we will need to have something more than 4 pictographs, i was thinking this 4 emoji would be represnting first 4 characters and last 4 characters of the Hex Address (plain deterministic encoding as Momcode), but brute forcing 4 bytes can be done with vanity generator with reasonably little time.. 

So we may need longer than 4 pictos, something at least > 6 bytes to be considered secure..

**Comment by gtlewis** [comment]
Authorisation requests should expire (I e. dropped from the UI) after a period of time - say 15 minutes. So the question is how many attempts to create the same symbols can be made in 15 minutes. If 
we think, say, 100million is enough combinations then 4 symbols with 100 in the set is adequate.

**Comment by alexvandesande** [comment]
There are two threat models that we need to address:

1. A malicious relayer sees a request and changes it to their own address
2. The relayer is honest, but an attacker is sending multiple auth requests trying to trick someone in authorizing the wrong one

To address (1) I would go with @drhus approach of showing 12 symbols, 4 correct and 8 random, and asks the user to put them in order. This is simply to guarantee that the key being added is the one they created on the other end. The requester shows the symbols, the authorizer inputs them.

To address (2) we could add an "emoji pin" in which we would invert the action and request the requester to type the pin in order to be able to make a request. This pin could be always the same, and the person can memorize (they can be reminded of it on their app) or it could be generated on the fly. 

I like the solution to (2) but I feel that it adds a lot of complexity (although I do believe users are used to typing emojis so finding them won't be a burden) in the whole interaction for a problem that simply doesn't exist anymore and can be first addressed by rate limiting, or putting a captcha on the requester window.

Basically: I would add (1) as a solution first (but it's not urgent) and would consider (2) if other more basic spam protection fails, as I rather not need to add a pin right now.

**Comment by alexvandesande** [comment]
I would however add more than just the 256 selection you have shown. With that restricted set, 4 symbols represent the first 8 chars of an address, and that can be mined in a few seconds, while if we increase the symbol set to 1024, then 4 emojis represent 10 characters, which would take years in an average computer. We can even increase the security by picking 10 random characters from the address, instead of the first one. That almost for free, since the end user still only needs to pick the characters from a set of 12 symbols. Here's what a set of 32x32 symbols looks like (there are even more emoji than that, so we could even replace a few of the weirder ones) 

<img width="706" alt="screen shot 2019-01-22 at 9 49 18 am" src="https://user-images.githubusercontent.com/112898/51534061-d48dca80-1e2b-11e9-9285-403a52e67be7.png">

**Comment by marekkirejczyk** [comment]
Implemented!
